packages/@aws-cdk/aws-ec2/lib/cfn-init-elements.ts

```diff
@@ -0,0 +1,395 @@
+
+/**
+ * The platform to which the init template applies.
+ */
+export enum InitPlatform {
+  /**
+   * Render the config for a Windows platform
+   */
+  WINDOWS = 'WINDOWS',
+
+  /**
+   * Render the config for a Linux platform
+   */
+  LINUX = 'LINUX',
+}
+
+/**
+ * The type of the init element.
+ */
+export enum InitElementType {
+  /**
+   * A package
+   */
+  PACKAGE = 'PACKAGE',
+
+  /**
+   * A group
+   */
+  GROUP = 'GROUP',
+
+  /**
+   * A user
+   */
+  USER = 'USER',
+
+  /**
+   * A source
+   */
+  SOURCE = 'SOURCE',
+
+  /**
+   * A file
+   */
+  FILE = 'FILE',
+
+  /**
+   * A command
+   */
+  COMMAND = 'COMMAND',
+
+  /**
+   * A service
+   */
+  SERVICE = 'SERVICE',
+}
+
+/**
+ * An object that represents reasons to restart an InitService
+ *
+ * Pass an instance of this object to all the `InitFile`, `InitCommand`,
+ * `InitSource` and `InitPackage` objects that you want to restart
+ * a service, and finally to the `InitService` itself as well.
+ */
+export class InitServiceRestartHandle {
+  private readonly commands = new Array<string>();
+  private readonly files = new Array<string>();
+  private readonly sources = new Array<string>();
+  private readonly packages: Record<string, string[]> = {};
+
+  /**
+   * Add a command key to the restart set
+   * @internal
+   */
+  public _addCommand(key: string) {
+    return this.commands.push(key);
+  }
+
+  /**
+   * Add a file key to the restart set
+   * @internal
+   */
+  public _addFile(key: string) {
+    return this.files.push(key);
+  }
+
+  /**
+   * Add a source key to the restart set
+   * @internal
+   */
+  public _addSource(key: string) {
+    return this.sources.push(key);
+  }
+
+  /**
+   * Add a package key to the restart set
+   * @internal
+   */
+  public _addPackage(packageType: string, key: string) {
+    if (!this.packages[packageType]) {
+      this.packages[packageType] = [];
+    }
+    this.packages[packageType].push(key);
+  }
+
+  /**
+   * Render the restart handles for use in an InitService declaration
+   * @internal
+   */
+  public _renderRestartHandles(): any {
+    const nonEmpty = <A>(x: A[]) => x.length > 0 ? x : undefined;
+
+    return {
+      commands: nonEmpty(this.commands),
+      files: nonEmpty(this.files),
+      packages: Object.keys(this.packages).length > 0 ? this.packages : undefined,
+      sources: nonEmpty(this.sources),
+    };
+  }
+}
+
+/**
+ * Base class for all CloudFormation Init elements
+ */
+export abstract class InitElement {
+
+  /**
+   * Returns the init element type for this element.
+   */
+  public abstract readonly elementType: InitElementType;
+
+  /**
+   * Called when the Init config is being consumed. Renders the CloudFormation
+   * representation of this init element, and calculates any authentication
+   * properties needed, if any.
+   *
+   * @param options bind options for the element.
+   * @internal
+   */
+  public abstract _bind(options: InitBindOptions): InitElementConfig;
```

Comment:
Would be nice to add a comment somewhere around here that this `InitElementConfig` exposes the underlying CloudFormation spec and should be fixed before making this API public. So as to inform someone in the future looking to open this up.